### ISSUE TITLE
Any way to remove unresolved asynchronous dependencies?

### ISSUE THREAD

**Issue by dmtaub** [pcafstockf/async-injection#19]
If we wanted to mark a dependency as no longer valid, and then downstream dependencies that rely on it and are awaiting could throw and be caught, is there a way to do that in this library?

**Comment by pcafstockf** [comment]
If you were to [removeBinding](https://pcafstockf.github.io/async-injection/api-docs/container.html#removebinding) of an upstream element, and then try to create the downstream dependent, the library would not find a registration for the upstream element and yes, it would throw. 

Is that what you were asking?

If you were asking about aborting a hanging promise (async factory or postconstruct method), I am not aware of any way to do that in JavaScript until the promise chooses to resolve or reject.

**Comment by dmtaub** [comment]
Yes, I was talking about the latter -- probably would require a timeout and retry cycle so that we can re-check that the binding exists between retries... We would wrap the entire thing in a promise so it would end up rejecting if the key became unavailable.

**Comment by pcafstockf** [comment]
Seems like that would incur a runtime penalty for the majority of the user base, for an unusual feature that can just as easily be implemented in user-land.
Whatever code is hanging could just as easily implement a timeout/retry cycle and throw if it hits its timeout.

Thank you for the feedback.  I feel this would be outside the core scope of the library.

-Frank

**Comment by dmtaub** [comment]
Yes, I intend to do the timeout and the throw for the dependencies myself, but the piece that I wasn't sure the library had was a way to handle the error during injection. I have looked into the test cases and now see we can do `.onError` from the `bindAsyncFactory` which if I understand correctly will work during *binding* though not during *injection*. Am I missing something?

I am imagining something like `@OnErrorFallback` to complement `@Optional` or perhaps a `@RejectHandler` to complement `@PostConstruct` -- here's a failing test case for the former, which assumes that the existing `@Optional` decorator handles the case of rejections.

```
	it('Async initialization optionally depending on an Async dependency should succeed', async () => {
		@Injectable()
		class A {
			public constructor() {
				this.a = 'A';
			}

			public a: string;

			@PostConstruct()
			public init(): Promise<void> {
				return new Promise<void>((resolve) => {
					setTimeout(() => {
						resolve();
					}, 25);
				});
			}
		}

		@Injectable()
		class B {
			public constructor() {
				this.b = 'B'
			}
			public b: string;

			@PostConstruct()
			public init(): Promise<void> {
				return new Promise<void>((resolve, reject) => {
					setTimeout(() => {
						reject(new Error('Failed post construction of B'));
					}, 25);
				});
			}
		}

		@Injectable()
		class C {
			public constructor(@Inject(B) @Optional({b: 'fallback'}) public b: {b: string}) {
				this.c = 'C'
				this.b = b;
			}
			public c: string;
		}

		const container = new Container();
		container.bindClass(A);
		container.bindClass(B);
		container.bindClass(C);
		let c = await container.resolve(C);
		expect(c.b.b).toBe('fallback');

	});
```

**Comment by pcafstockf** [comment]
Thank you! 
I have added your sample as a test case since that should work.

I did trace through it enough to realize this is not a quick / easy fix, but I will prioritize it and get a new version out asap.

Also, thank you for the hint...  My first reaction is the root problem may be that the `.onError` from the `bindAsyncFactory` is **not** contemplating async binding errors but **should** be.

**Comment by pcafstockf** [comment]
Apologies for the delay in tracking this one down, I've been buried trying to get [another project](https://github.com/pcafstockf/oag-tools) out.

I appreciate your mention of `@Optional`, as the documentation for that explicitly says its value will be used "if for **any** reason it is unable to otherwise resolve the type of the parameter".
